File: trie/trie.ts
class TrieNode {
    children: Map<string, TrieNode> = new Map();
    isEnd: boolean = false;
}

class Trie {
    private root = new TrieNode();
    
    insert(word: string): void {
        let node = this.root;
        for (const char of word) {
            if (!node.children.has(char)) {
                node.children.set(char, new TrieNode());
            }
            node = node.children.get(char)!;
        }
        node.isEnd = true;
    }
    
    search(word: string): boolean {
        let node = this.root;
        for (const char of word) {
            if (!node.children.has(char)) return false;
            node = node.children.get(char)!;
        }
        return node.isEnd;
    }
}

const trie = new Trie();
trie.insert("hello");
console.log(trie.search("hello"));

File: trie/Trie.kt
class TrieNode {
    val children = mutableMapOf<Char, TrieNode>()
    var isEnd = false
}

class Trie {
    private val root = TrieNode()
    
    fun insert(word: String) {
        var node = root
        for (char in word) {
            node = node.children.getOrPut(char) { TrieNode() }
        }
        node.isEnd = true
    }
    
    fun search(word: String): Boolean {
        var node = root
        for (char in word) {
            node = node.children[char] ?: return false
        }
        return node.isEnd
    }
    
    fun startsWith(prefix: String): Boolean {
        var node = root
        for (char in prefix) {
            node = node.children[char] ?: return false
        }
        return true
    }
}

fun main() {
    val trie = Trie()
    trie.insert("hello")
    trie.insert("world")
    println("Search 'hello': ${trie.search("hello")}")
    println("Starts with 'hel': ${trie.startsWith("hel")}")
}
